`nice_name` should be used in diagnostics of `extern crate icky_name as nice_name`

### Code

```Rust
extern crate unwieldy_crate_name as nice_crate_name;

fn use_foo_from_another_crate_without_importing_it_first() {
    // This will expectedly error out saying:
    // error[E0412]: cannot find type `Foo` in this scope
    //
    // The ergonomics issue is that the fix suggestion may refer to the
    // `unwieldy_crate_name` instead of `nice_crate_name`:
    //
    // help: consider importing this struct
    //   |
    // 1 + use unwieldy_crate_name::Foo;
    //   |

    let _: Foo<_> = todo!();
}
```


### Current output

```Shell
error[E0412]: cannot find type `Foo` in this scope
  --> ./repro.rs:15:12
   |
15 |     let _: Foo<_> = todo!();
   |            ^^^ not found in this scope
   |
help: consider importing this struct
   |
1  + use unwieldy_crate_name::Foo;
   |
```


### Desired output

```Shell
error[E0412]: cannot find type `Foo` in this scope
  --> ./repro.rs:15:12
   |
15 |     let _: Foo<_> = todo!();
   |            ^^^ not found in this scope
   |
help: consider importing this struct
   |
1  + use nice_crate_name::Foo;
```


### Rationale and extra context

Some build environments (e.g. Chromium (*), Bazel) may generate crate names based on names of build targets.  While this results in globally-unique (**) crate names, it results in long and quite unwieldy crate names.  Both Chromium and Blaze provide a macro to help working with such crate names:

```
chromium::import! {
    "//third_party/blink/renderer/platform/image-decoders/rust_bindings:scoped_refptr";
}
```

The macro usage above expands to something like:

```
extern crate third_uparty_sblink_srenderer_splatform_simage_ddecoders_srust_ubindings_cscoped_urefptr as scoped_refptr; 
```

Unfortunately, it seems that some `rustc` diagnostics keep referring to the long and unwieldy name.

(*) see [`chromium_prelude/import_attribute.rs`](https://source.chromium.org/chromium/chromium/src/+/main:build/rust/chromium_prelude/import_attribute.rs;drc=52ee534619c5389ecfcb93bfb5513bf176b42dc4) and [`chromium/build/rust/rust_static_library.gni`](https://source.chromium.org/chromium/chromium/src/+/main:build/rust/rust_static_library.gni;l=175-199;drc=52ee534619c5389ecfcb93bfb5513bf176b42dc4)
(**) globally-unique crate names = within a given build tree

### Other cases

_No response_

### Rust Version

```Shell
$ rustc --version --verbose
rustc 1.78.0-nightly (ee9c7c940 2024-02-14)
binary: rustc
commit-hash: ee9c7c940c07d8b67c9a6b2ec930db70dcd23a46
commit-date: 2024-02-14
host: x86_64-unknown-linux-gnu
release: 1.78.0-nightly
LLVM version: 18.1.0
```


### Anything else?

More detailed repro can be found below:

```
$ cat unwieldy_crate_name.rs
pub struct Foo<T>(pub core::ptr::NonNull<T>);

$ cat repro.rs
extern crate unwieldy_crate_name as nice_crate_name;

fn use_foo_from_another_crate_without_importing_it_first() {
    // This will expectedly error out saying:
    // error[E0412]: cannot find type `Foo` in this scope
    //
    // The ergonomics issue is that the fix suggestion may refer to the
    // `unwieldy_crate_name` instead of `nice_crate_name`:
    //
    // help: consider importing this struct
    //   |
    // 1 + use unwieldy_crate_name::Foo;
    //   |

    let _: Foo<_> = todo!();
}

$ rustc --crate-type rlib --edition=2021 ./unwieldy_crate_name.rs

$ rustc --crate-type rlib --edition=2021 ./repro.rs --extern unwieldy_crate_name=./libunwieldy_crate_name.rlib
error[E0412]: cannot find type `Foo` in this scope
  --> ./repro.rs:15:12
   |
15 |     let _: Foo<_> = todo!();
   |            ^^^ not found in this scope
   |
help: consider importing this struct
   |
1  + use unwieldy_crate_name::Foo;
   |

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0412`.
```

BTW: I also want to note/highlight that I am not able to repro the problem without using `--edition=2021`.  For example (same `.rs` files):

```
$ rustc --crate-type rlib ./unwieldy_crate_name.rs

$ rustc --crate-type rlib ./repro.rs --extern unwieldy_crate_name=./libunwieldy_crate_name.rlib
error[E0412]: cannot find type `Foo` in this scope
  --> ./repro.rs:15:12
   |
15 |     let _: Foo<_> = todo!();
   |            ^^^ not found in this scope
   |
help: consider importing this struct
   |
1  + use nice_crate_name::Foo;
   |

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0412`.
```

--- a/compiler/rustc_resolve/src/diagnostics.rs
+++ b/compiler/rustc_resolve/src/diagnostics.rs
@@ -1290,10 +1290,20 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                     let mut path_segments = path_segments.clone();
                     path_segments.push(ast::PathSegment::from_ident(ident));
 
+                    let alias_import = if let NameBindingKind::Import { import, .. } =
+                        name_binding.kind
+                        && let ImportKind::ExternCrate { source: Some(_), .. } = import.kind
+                        && import.parent_scope.expansion == parent_scope.expansion
+                    {
+                        true
+                    } else {
+                        false
+                    };
+
                     let is_extern_crate_that_also_appears_in_prelude =
                         name_binding.is_extern_crate() && lookup_ident.span.at_least_rust_2018();
 
-                    if !is_extern_crate_that_also_appears_in_prelude {
+                    if !is_extern_crate_that_also_appears_in_prelude || alias_import {
                         // add the module to the lookup
                         if seen_modules.insert(module.def_id()) {
                             if via_import { &mut worklist_via_import } else { &mut worklist }

--- a/tests/ui/imports/auxiliary/import-alias-issue-121168-extern.rs
+++ b/tests/ui/imports/auxiliary/import-alias-issue-121168-extern.rs
@@ -1,0 +1,1 @@
+pub struct Foo<T>(pub core::ptr::NonNull<T>);

--- a/tests/ui/imports/import-alias-issue-121168.edition2015.stderr
+++ b/tests/ui/imports/import-alias-issue-121168.edition2015.stderr
@@ -1,0 +1,14 @@
+error[E0412]: cannot find type `Foo` in this scope
+  --> $DIR/import-alias-issue-121168.rs:11:12
+   |
+LL |     let _: Foo<i32> = todo!();
+   |            ^^^ not found in this scope
+   |
+help: consider importing this struct
+   |
+LL + use nice_crate_name::Foo;
+   |
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0412`.

--- a/tests/ui/imports/import-alias-issue-121168.edition2018.stderr
+++ b/tests/ui/imports/import-alias-issue-121168.edition2018.stderr
@@ -1,0 +1,16 @@
+error[E0412]: cannot find type `Foo` in this scope
+  --> $DIR/import-alias-issue-121168.rs:11:12
+   |
+LL |     let _: Foo<i32> = todo!();
+   |            ^^^ not found in this scope
+   |
+help: consider importing one of these items
+   |
+LL + use crate::nice_crate_name::Foo;
+   |
+LL + use import_alias_issue_121168_extern::Foo;
+   |
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0412`.

--- a/tests/ui/imports/import-alias-issue-121168.edition2021.stderr
+++ b/tests/ui/imports/import-alias-issue-121168.edition2021.stderr
@@ -1,0 +1,16 @@
+error[E0412]: cannot find type `Foo` in this scope
+  --> $DIR/import-alias-issue-121168.rs:11:12
+   |
+LL |     let _: Foo<i32> = todo!();
+   |            ^^^ not found in this scope
+   |
+help: consider importing one of these items
+   |
+LL + use crate::nice_crate_name::Foo;
+   |
+LL + use import_alias_issue_121168_extern::Foo;
+   |
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0412`.

--- a/tests/ui/imports/import-alias-issue-121168.rs
+++ b/tests/ui/imports/import-alias-issue-121168.rs
@@ -1,0 +1,14 @@
+//@ revisions: edition2015 edition2018 edition2021
+//@ [edition2015] edition:2015
+//@ [edition2018] edition:2018
+//@ [edition2021] edition:2021
+//@ compile-flags: --extern import_alias_issue_121168_extern
+//@ aux-build: import-alias-issue-121168-extern.rs
+
+extern crate import_alias_issue_121168_extern as nice_crate_name;
+
+fn use_foo_from_another_crate_without_importing_it_first() {
+    let _: Foo<i32> = todo!(); //~ ERROR cannot find type `Foo` in this scope
+}
+
+fn main() {}

